evidence/pool.go

```diff
@@ -177,35 +283,162 @@ func (evpool *Pool) MarkEvidenceAsCommitted(height int64, evidence []types.Evide
+	evpb, err := types.EvidenceToProto(ei.Evidence)
+	if err != nil {
+		return nil, err
+	}
+
+	valsProto := make([]*tmproto.Validator, len(ei.Validators))
+	for i := 0; i < len(ei.Validators); i++ {
+		valp, err := ei.Validators[i].ToProto()
+		if err != nil {
+			return nil, err
+		}
+		valsProto[i] = valp
+	}
+
+	return &evproto.Info{
+		Evidence:         *evpb,
+		Time:             ei.Time,
+		Validators:       valsProto,
+		TotalVotingPower: ei.TotalVotingPower,
+	}, nil
+}
+
+// InfoFromProto decodes from protobuf into Info
+func InfoFromProto(proto *evproto.Info) (Info, error) {
+	if proto == nil {
+		return Info{}, errors.New("nil evidence info")
+	}
+
+	ev, err := types.EvidenceFromProto(&proto.Evidence)
+	if err != nil {
+		return Info{}, err
+	}
+
+	vals := make([]*types.Validator, len(proto.Validators))
+	for i := 0; i < len(proto.Validators); i++ {
+		val, err := types.ValidatorFromProto(proto.Validators[i])
+		if err != nil {
+			return Info{}, err
+		}
+		vals[i] = val
+	}
+
+	return Info{
+		Evidence:         ev,
+		Time:             proto.Time,
+		Validators:       vals,
+		TotalVotingPower: proto.TotalVotingPower,
+	}, nil
+
+}
+
+//--------------------------------------------------------------------------
+
+// allPendingEvidence returns all evidence ready to be proposed and committed.
+func (evpool *Pool) allPendingEvidence() []types.Evidence {
+	evidence, err := evpool.listEvidence(baseKeyPending, -1)
+	if err != nil {
+		evpool.logger.Error("Unable to retrieve pending evidence", "err", err)
+	}
+	return evidence
+}
+
+// fastCheck leverages the fact that the evidence pool may have already verified the evidence to see if it can
+// quickly conclude that the evidence is already valid.
+func (evpool *Pool) fastCheck(ev types.Evidence) bool {
+	key := keyPending(ev)
+	if lcae, ok := ev.(*types.LightClientAttackEvidence); ok {
+		evBytes, err := evpool.evidenceStore.Get(key)
+		if evBytes == nil { // the evidence is not in the nodes pending list
+			return false
+		}
+		if err != nil {
+			evpool.logger.Error("Failed to get evidence during fastcheck", "err", err, "ev", lcae)
+			return false
+		}
+		var evpb *evproto.Info
+		err = evpb.Unmarshal(evBytes)
+		if err != nil {
+			evpool.logger.Error("Failed to unmarshal evidence info during fastcheck", "err", err)
+			return false
+		}
+		evInfo, err := InfoFromProto(evpb)
+		if err != nil {
+			evpool.logger.Error("Failed to convert evidence from proto during fastcheck", "err", err)
+			return false
+		}
+		// ensure that all the validators that the evidence pool have found to be malicious
+		// are present in the list of commit signatures in the conflicting block
+	OUTER:
+		for _, sig := range lcae.ConflictingBlock.Commit.Signatures {
+			for _, val := range evInfo.Validators {
+				if bytes.Equal(val.Address, sig.ValidatorAddress) {
+					continue OUTER
+				}
+			}
+			// a validator we know is malicious is not included in the commit
```

Comment:
isn't it vice versa: `a signature from conflicting block is not present in validator set` ?

Reply:
hmm I think it could be both. I was thinking of the case where some replaces commit Sigs with absent commits in which case we have more validators that we found to be malicious than what this piece of evidence suggests

Reply:
Basically the case I have in mind is that a perfectly good lightclientattackevidence is created and gossiped then someone proposes a lightclientattackevidence to the block with the exact same conflicting header but they have altered the commit such that they are all absent signatures. When people are verifying the evidence they will see that the hashes are the same because the headers are the same and they will therefore vote on the evidence where the commit is really empty and when it comes to finding the malicious validators none will be present. Hence this check makes sure that the validators I know are bad matches the validators that this piece of evidence claims is bad